acvm-repo/brillig/src/opcodes.rs

```diff
@@ -19,18 +19,41 @@ impl From<usize> for RegisterIndex {
     }
 }
 
+/// Describes the memory layout for an array/vector element
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum HeapValueType {
+    // A single field element is enough to represent the value
+    Simple,
+    // The value read should be interpreted as a pointer to a heap array, which
+    // consists of a pointer to a slice of memory of size elements, and a
+    // reference count
+    Array { value_types: Vec<HeapValueType>, size: usize },
+    // The value read should be interpreted as a pointer to a heap vector, which
+    // consists of a pointer to a slice of memory, a number of elements in that
+    // slice, and a reference count
+    Vector { value_types: Vec<HeapValueType> },
+}
+
+impl HeapValueType {
+    pub fn all_simple(types: &[HeapValueType]) -> bool {
+        types.iter().all(|typ| matches!(typ, HeapValueType::Simple))
+    }
+}
+
 /// A fixed-sized array starting from a Brillig register memory location.
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Copy)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct HeapArray {
     pub pointer: RegisterIndex,
     pub size: usize,
+    pub value_types: Vec<HeapValueType>,
```

Comment:
One general question regarding the approach: could it be possible for this value_types to be held in the ForeignCall opcode instead of the RegisterOrMemory structs? RegisterOrMemory types are used for non-foreign-call-related operations that don't need the typing information, such as black box opcodes

I'll take a deeper look next week!

Reply:
Hmm... that's a good point. Black box operations don't need them because the types they accept are known at compile time. Yes, I think it should be possible to put the `value_types` in the `ForeignCall` opcode. That would reduce the size of the PR and the surface of the change. It would still change the serialization format though :/ I will try to make the change and update the PR.

Reply:
Thanks a lot! This is looking a much smaller change now. Will do a more in-depth review!